pkg/lease/manager.go

```diff
@@ -0,0 +1,235 @@
+		if setAcquireAndLeaseTransitions {
+			lease.Spec.AcquireTime = &currentTime
+			if lease.Spec.LeaseTransitions != nil {
+				*lease.Spec.LeaseTransitions += int32(1)
+			} else {
+				lease.Spec.LeaseTransitions = pointer.Int32(1)
+			}
+		}
+		owner := makeExpectedOwnerOfLease(obj)
+		lease.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*owner}
+		lease.Spec.HolderIdentity = &l.holderIdentity
+		lease.Spec.LeaseDurationSeconds = pointer.Int32(int32(leaseDuration.Seconds()))
+		lease.Spec.RenewTime = &currentTime
+		if err := l.Client.Update(ctx, lease); err != nil {
+			log.Errorf("Failed to update the lease. obj %s error: %v", obj.GetName(), err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (l *manager) invalidateLease(ctx context.Context, obj client.Object) error {
+	log.Info("invalidating lease")
+	lease, err := l.getLease(ctx, obj)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	if err := l.Client.Delete(ctx, lease); err != nil {
+		log.Error(err, "failed to delete lease to be invalidated")
+		return err
+	}
+
+	return nil
+}
+
+func makeExpectedOwnerOfLease(obj client.Object) *metav1.OwnerReference {
+
+	return &metav1.OwnerReference{
+		APIVersion: obj.GetObjectKind().GroupVersionKind().Version,
+		Kind:       obj.GetObjectKind().GroupVersionKind().Kind,
+		Name:       obj.GetName(),
+		UID:        obj.GetUID(),
+	}
+}
+
+func leaseDueTime(lease *coordv1.Lease) time.Time {
+	return lease.Spec.RenewTime.Time.Add(time.Duration(*lease.Spec.LeaseDurationSeconds) * time.Second)
+}
+
+func needUpdateOwnedLease(lease *coordv1.Lease, currentTime metav1.MicroTime, requestedLeaseDuration time.Duration) (bool, bool) {
+
+	if lease.Spec.RenewTime == nil || lease.Spec.LeaseDurationSeconds == nil {
+		log.Info("empty renew time or duration in sec")
+		return true, true
+	}
+	dueTime := leaseDueTime(lease)
+
+	// if lease expired right now, then both update the lease and the acquire time (second rvalue)
+	// if the acquire time has been previously nil
+	if dueTime.Before(currentTime.Time) {
+		return true, lease.Spec.AcquireTime == nil
+	}
+
+	deadline := currentTime.Add(requestedLeaseDuration)
+
+	// about to expire, update the lease but not the acquired time (second value)
+	return dueTime.Before(deadline), false
+}
+
+func isValidLease(lease *coordv1.Lease, currentTime time.Time) bool {
+
+	if lease.Spec.RenewTime == nil || lease.Spec.LeaseDurationSeconds == nil {
+		return false
+	}
+
+	renewTime := (*lease.Spec.RenewTime).Time
+	dueTime := leaseDueTime(lease)
+
+	// valid lease if: due time not in the past and renew time not in the future
+	return !dueTime.Before(currentTime) && !renewTime.After(currentTime)
+}
+
+func (l *manager) getLease(ctx context.Context, obj client.Object) (*coordv1.Lease, error) {
+	log.Info("getting lease")
+	nName := apitypes.NamespacedName{Namespace: l.namespace, Name: obj.GetName()}
+	lease := &coordv1.Lease{}
+
+	getOption := &metav1.GetOptions{
```

Comment:
What do we need this for? 🤔

Reply:
oooohhhhh..... I missed that in the create function.

`TypeMeta` is about the type which is being created or read, so the Lease! Not the type of the object we create the Lease for. Basically you are using the Go `Lease` type now, but override the type internally with something different. Very interesting that it doesn't break the unit tests...

> we'd like to be able to distinct between a lease called abc where client.Object is a node  and a lease called abc where client.Object is a pod

My idea was to add the kind to the name of the lease as prefix. Using a label could also be an idea.

Reply:
Hmm I do have a question:
Do we want to add the `Kind / Version` of client.Object in order to make sure lease key is unique ?
In case the answer is yes, I think your suggestion of adding it as a prefix to the name makes sense, but I don't think it'll apply for using a label.

Or maybe we just want to make sure we are getting the correct lease when fetching ? 
In that case maybe we can just use the OwnerReferences which already has the Kind/Version (and than we don't need to add labels or manipulate the name).

WDYT ?